src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,974 @@
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(momGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 4\n");
+        // Add to neuron
+        idxInChild = childGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        idxInParent = momGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(momGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 5\n");
+        continue;
+      }
+
+      if (linkContainedInDad && randNum < 0.5) {
+        childGenome.AddLink(momGenome.aLinkGenes[i]);
+////printf("crossover 6\n");
+        // Add from neuron
+        ssize_t idxInChild = childGenome.GetNeuronIndex(momGenome.aLinkGenes[i].FromNeuronId());
+        ssize_t idxInParent = momGenome.GetNeuronIndex(momGenome.aLinkGenes[i].FromNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(momGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 7\n");
+        // Add to neuron
+        idxInChild = childGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        idxInParent = momGenome.GetNeuronIndex(momGenome.aLinkGenes[i].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(momGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 8\n");
+        continue;
+      }
+////printf("crossover 9\n");
+      if (linkContainedInDad && randNum >= 0.5) {
+        childGenome.AddLink(dadGenome.aLinkGenes[idx]);
+////printf("crossover 10\n");
+        // Add from neuron   TODO: make it a function?? check whether crossover is correct.
+        ssize_t idxInChild = childGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].FromNeuronId());
+        ssize_t idxInParent = dadGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].FromNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 11\n");
+        // Add to neuron
+        idxInChild = childGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        idxInParent = dadGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 12\n");
+        continue;
+      }  
+    }
+  }
+
+  // Crossover two genome to get one genome.
+  void Crossover(Genome& genome1, Genome& genome2, Genome& childGenome) {
+    if (CompareGenome(genome1, genome2)) {  // genome1 is better
+      CrossoverLinkAndNeuron(genome1, genome2, childGenome);
+    } else {
+      CrossoverLinkAndNeuron(genome2, genome1, childGenome);
+    }
+  }
+
+  // Measure two genomes' disjoint (including exceed).
+  // NOTICE: we can separate into disjoint and exceed. But currently maybe it is enough.
+  double Disjoint(Genome& genome1, Genome& genome2) {
```

Comment:
Let us distinguish between disjoint and excess here. At the same time, we can decrease the runtime, by not looping over the genomes twice:

```
size_t disjoint = 0;
size_t excess = 0;

if (genome1.NumLink() >= genome2.NumLink())
{
    DisjointExcess(genome1, genome2, disjoint, excess);
}
else
{
    DisjointExcess(genome2, genome1, disjoint, excess);
}


void DisjointExcess(Genome& genome1, Genome& genome2, size_t& disjoint, size_t& excess, size_t& matching)
{
    size_t maxInovation = genome2.GetMaxInnovation();

    for (size_t i = 0; i < genome1.NumLink(); ++i)
    {
        size_t innovationNumber = genome1.aLinkGenes[i].InnovationId();

        if (!genome2.ContainLink(innovId))
        {
            if (innovationNumber > maxInovation)
            {
                excess += 1;
            }
            else
            {
                disjoint += 1;
            }
        }
        else
        {
            matching += 1;
        }
    }

    disjoint += genome2.NumLink() - matching;
}
```